File: internal/status/status.go
package status

import (
	"bufio"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strings"
	"time"

	"github.com/delightfulhammers/telesis/internal/config"
)

// Status holds the aggregated project state.
type Status struct {
	ProjectName        string
	ProjectStatus      string
	ADRCount           int
	TDDCount           int
	ActiveMilestone    string
	ContextGeneratedAt time.Time
}

// GetStatus reads the project state from the filesystem.
func GetStatus(rootDir string) (*Status, error) {
	cfg, err := config.Load(rootDir)
	if err != nil {
		return nil, err
	}

	adrCount, err := countFiles(filepath.Join(rootDir, "docs", "adr"), "ADR-*.md")
	if err != nil {
		return nil, fmt.Errorf("counting ADRs: %w", err)
	}

	tddCount, err := countFiles(filepath.Join(rootDir, "docs", "tdd"), "TDD-*.md")
	if err != nil {
		return nil, fmt.Errorf("counting TDDs: %w", err)
	}

	milestone, err := extractActiveMilestone(filepath.Join(rootDir, "docs", "MILESTONES.md"))
	if err != nil {
		return nil, fmt.Errorf("reading milestones: %w", err)
	}

	contextTime := contextTimestamp(filepath.Join(rootDir, "CLAUDE.md"))

	return &Status{
		ProjectName:        cfg.Project.Name,
		ProjectStatus:      cfg.Project.Status,
		ADRCount:           adrCount,
		TDDCount:           tddCount,
		ActiveMilestone:    milestone,
		ContextGeneratedAt: contextTime,
	}, nil
}

func countFiles(dir, pattern string) (int, error) {
	entries, err := os.ReadDir(dir)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return 0, nil
		}
		return 0, err
	}

	count := 0
	for _, entry := range entries {
		if entry.IsDir() {
			continue
		}
		matched, _ := filepath.Match(pattern, entry.Name())
		if matched {
			count++
		}
	}
	return count, nil
}

var milestoneHeaderRe = regexp.MustCompile(`(?i)^##\s+MVP`)

func extractActiveMilestone(path string) (string, error) {
	f, err := os.Open(path)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return "", nil
		}
		return "", err
	}
	defer f.Close()

	var capturing bool
	var lines []string

	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		line := scanner.Text()
		if milestoneHeaderRe.MatchString(line) {
			capturing = true
			lines = append(lines, line)
			continue
		}
		if capturing {
			if strings.HasPrefix(line, "## ") || line == "---" {
				break
			}
			lines = append(lines, line)
		}
	}
	if err := scanner.Err(); err != nil {
		return "", fmt.Errorf("scanning milestones: %w", err)
	}

	return strings.TrimSpace(strings.Join(lines, "\n")), nil
}

func contextTimestamp(path string) time.Time {
	info, err := os.Stat(path)
	if err != nil {
		return time.Time{}
	}
	return info.ModTime()
}

File: src/status/status.ts
import { readdirSync, statSync as fsStatSync, type Dirent } from "node:fs";
import { join, resolve } from "node:path";
import { load } from "../config/config.js";
import { extractActiveMilestone } from "../milestones/parse.js";
import { streamTelemetryRecords } from "../agent/telemetry/reader.js";
import {
  loadPricing,
  costForRecord,
  type PricingConfig,
} from "../agent/telemetry/pricing.js";
import type { ModelCallRecord } from "../agent/telemetry/types.js";
import { loadNotes } from "../notes/store.js";

export interface Status {
  readonly projectName: string;
  readonly projectStatus: string;
  readonly adrCount: number;
  readonly tddCount: number;
  readonly activeMilestone: string;
  readonly contextGeneratedAt: Date | null;
  readonly totalInputTokens: number;
  readonly totalOutputTokens: number;
  readonly modelCallCount: number;
  readonly noteCount: number;
  readonly estimatedCost: number | null;
}

const countFiles = (dir: string, pattern: RegExp): number => {
  let entries: Dirent[];
  try {
    entries = readdirSync(dir, { withFileTypes: true });
  } catch (err) {
    if ((err as NodeJS.ErrnoException).code === "ENOENT") return 0;
    throw err;
  }

  return entries.filter(
    (entry) => !entry.isDirectory() && pattern.test(entry.name),
  ).length;
};

const contextTimestamp = (path: string): Date | null => {
  try {
    const info = fsStatSync(path);
    return info.mtime;
  } catch {
    return null;
  }
};

export const getStatus = async (rootDir: string): Promise<Status> => {
  const cfg = load(rootDir);

  const adrCount = countFiles(join(rootDir, "docs", "adr"), /^ADR-.*\.md$/);

  const tddCount = countFiles(join(rootDir, "docs", "tdd"), /^TDD-.*\.md$/);

  const activeMilestone = extractActiveMilestone(
    join(rootDir, "docs", "MILESTONES.md"),
  );

  const contextGeneratedAt = contextTimestamp(join(rootDir, "CLAUDE.md"));

  const telemetryPath = join(resolve(rootDir), ".telesis", "telemetry.jsonl");
  let totalInputTokens = 0;
  let totalOutputTokens = 0;
  let modelCallCount = 0;

  const pricing: PricingConfig | null = loadPricing(rootDir);

  const recordCost = (record: ModelCallRecord): number => {
    if (!pricing) return 0;
    const mp = pricing.models[record.provider]?.[record.model];
    return mp ? costForRecord(record, mp) : 0;
  };

  let totalCost = 0;
  for await (const record of streamTelemetryRecords(telemetryPath)) {
    totalInputTokens += record.inputTokens;
    totalOutputTokens += record.outputTokens;
    totalCost += recordCost(record);
    modelCallCount++;
  }

  const noteCount = loadNotes(rootDir).items.length;

  return {
    projectName: cfg.project.name,
    projectStatus: cfg.project.status,
    adrCount,
    tddCount,
    activeMilestone,
    contextGeneratedAt,
    totalInputTokens,
    totalOutputTokens,
    noteCount,
    modelCallCount,
    estimatedCost: modelCallCount > 0 && pricing ? totalCost : null,
  };
};
